Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc

```diff
@@ -0,0 +1,177 @@
+
+      // create new node with the same information as in sDC and add it as a child of pD
+      DetGeomDesc* cD = new DetGeomDesc(*sDC);
+      pD->addComponent(cD);
+
+      bufferNew.emplace_back(cD);
+    }
+  }
+}
+
+std::unique_ptr<DetGeomDesc> PPSGeometryESProducer::produceIdealGD(const IdealGeometryRecord& iRecord) {
+  // Get the DDDetector from EventSetup
+  auto const& det = iRecord.get(detectorToken_);
+
+  // Get the DDCompactView
+  cms::DDCompactView myCompactView(det);
+
+  // Build geo from compact view.
+  return DetGeomDescBuilder::buildDetGeomDescFromCompactView(myCompactView);
+}
+
+template <typename REC>
+std::unique_ptr<DetGeomDesc> PPSGeometryESProducer::produceGD(IdealGeometryRecord const& iIdealRec,
+                                                              std::optional<REC> const& iAlignRec,
+                                                              GDTokens<REC> const& iTokens,
+                                                              const char* name) {
+  // get the input GeometricalDet
+  auto const& idealGD = iIdealRec.get(iTokens.idealGDToken_);
+
+  // load alignments
+  edm::ESHandle<CTPPSRPAlignmentCorrectionsData> alignments;
+  if (iAlignRec) {
+    alignments = iAlignRec->getHandle(iTokens.alignmentToken_);
+  }
+
+  if (alignments.isValid()) {
+    if (verbosity_)
+      edm::LogVerbatim(name) << ">> " << name << " > Real geometry: " << alignments->getRPMap().size() << " RP and "
+                             << alignments->getSensorMap().size() << " sensor alignments applied.";
+  } else {
+    if (verbosity_)
+      edm::LogVerbatim(name) << ">> " << name << " > Real geometry: No alignments applied.";
+  }
+
+  DetGeomDesc* newGD = nullptr;
+  applyAlignments(idealGD, alignments.product(), newGD);
+  return std::unique_ptr<DetGeomDesc>(newGD);
+}
+
+std::unique_ptr<DetGeomDesc> PPSGeometryESProducer::produceRealGD(const VeryForwardRealGeometryRecord& iRecord) {
+  return produceGD(iRecord.getRecord<IdealGeometryRecord>(),
+                   iRecord.tryToGetRecord<RPRealAlignmentRecord>(),
+                   gdRealTokens_,
+                   "PPSGeometryESProducer::produceRealGD");
+}
+
+//----------------------------------------------------------------------------------------------------
+
+std::unique_ptr<DetGeomDesc> PPSGeometryESProducer::produceMisalignedGD(
+    const VeryForwardMisalignedGeometryRecord& iRecord) {
+  return produceGD(iRecord.getRecord<IdealGeometryRecord>(),
+                   iRecord.tryToGetRecord<RPMisalignedAlignmentRecord>(),
+                   gdMisTokens_,
+                   "PPSGeometryESProducer::produceMisalignedGD");
+}
+
+//----------------------------------------------------------------------------------------------------
+
+std::unique_ptr<CTPPSGeometry> PPSGeometryESProducer::produceRealTG(const VeryForwardRealGeometryRecord& iRecord) {
+  auto const& gD = iRecord.get(dgdRealToken_);
+
+  return std::make_unique<CTPPSGeometry>(&gD, verbosity_);
```

Comment:
This will lead to double-delete. If an ESProduct is intended to be forwarded as such, it should be done as
```cpp
  return std::shared<CTPPSGeometry const>(&gD, edm::do_nothing_deleter());
```

Reply:
@ghugo83 Doesn't this double-delete have to fixed to have a working program? Wouldn't this line introduce a crashing bug into the master?

Reply:
As discussed, the thing is that this is already in master. This PR does not change this code.

Reply:
@ghugo83 I am not understanding. Matti said this line creates a double-delete, which produces undefined behavior. You are saying this undefined behavior is already in the master. If that is so, doesn't that undefined behavior have to be fixed as a top priority? What is a higher priority?
I am not suggesting an overall re-design of the memory management, just a quick fix to eliminate random behavior.